mqtt/mqtt-bridge/src/bridge.rs

```diff
@@ -1,491 +1,128 @@
-        let key = if let Some(local) = topic.local() {
-            format!("{}/{}", local, topic.pattern().to_string())
-        } else {
-            topic.pattern().into()
-        };
-        (key, topic.clone())
-    }
-
-    pub async fn start(&self) -> Result<(), BridgeError> {
-        info!("Starting bridge...{}", self.connection_settings.name());
+    ) -> Result<Self, BridgeError> {
+        debug!("creating bridge...{}", connection_settings.name());
 
-        self.connect_to_local().await?;
-        self.connect_to_remote().await?;
-
-        Ok(())
-    }
-
-    async fn connect_to_remote(&self) -> Result<(), BridgeError> {
-        info!(
-            "connecting to remote broker {}",
-            self.connection_settings.address()
-        );
-
-        self.connect(
-            self.subscriptions.clone(),
-            self.connection_settings.address(),
-            self.connection_settings.credentials(),
-            true,
-        )
-        .await
-    }
-
-    async fn connect_to_local(&self) -> Result<(), BridgeError> {
-        let client_id = format!(
+        let local_client_id = format!(
             "{}/$edgeHub/$bridge/{}",
-            self.device_id,
-            self.connection_settings.name()
-        );
-        info!(
-            "connecting to local broker {}, clientid {}",
-            self.system_address, client_id
+            device_id,
+            connection_settings.name()
         );
 
-        self.connect(
-            self.forwards.clone(),
-            self.system_address.as_str(),
-            &Credentials::Anonymous(client_id),
-            false,
-        )
-        .await
-    }
-
-    async fn connect(
-        &self,
-        mut topics: HashMap<String, Topic>,
-        address: &str,
-        credentials: &Credentials,
-        secure: bool,
-    ) -> Result<(), BridgeError> {
-        let (subscriptions, topics): (Vec<_>, Vec<_>) = topics.drain().unzip();
-        let topic_filters = topics
-            .into_iter()
-            .map(|topic| topic.try_into())
-            .collect::<Result<Vec<_>, _>>()?;
+        let mut pumps =
+            PumpPair::new(connection_settings.clone(), system_address, local_client_id)?;
 
-        let persistor = PublicationStore::new_memory(BATCH_SIZE);
-        let mut client = if secure {
-            MqttClient::tls(
-                address,
-                self.connection_settings.keep_alive(),
-                self.connection_settings.clean_session(),
-                MessageHandler::new(persistor, topic_filters),
-                credentials,
-            )
-        } else {
-            MqttClient::tcp(
-                address,
-                self.connection_settings.keep_alive(),
-                self.connection_settings.clean_session(),
-                MessageHandler::new(persistor, topic_filters),
-                credentials,
-            )
-        };
+        pumps.local_pump.subscribe().await?;
```

Comment:
I was thinking that when the subscription update change comes through, this can be changed then. I am not sure where is best to move it now. We could move it to run() now but it is going to be doing the same thing.

Reply:
makes sense I guess